tests/test_types_self.py

```diff
@@ -37,4 +37,39 @@ class SelfRef(BaseModel):
         ValidationError,
         match=r'ref\.ref\s+Input should be a valid dictionary or instance of SelfRef \[type=model_type,',
     ):
-        SelfRef(data=1, ref={'data': 2, 'ref': 3}).model_dump() == {'data': 1, 'ref': {'data': 2, 'ref': None}}
+        SelfRef(data=1, ref={'data': 2, 'ref': 3}).model_dump()
+
+
+def test_recursive_model_with_subclass(Self):
+    """
+    Self refs should be valid in covariant direction
+    """
+
+    class SelfRef(BaseModel):
+        data: int
+        ref: Self | None = None
+
+    class SubSelfRef(SelfRef):
+        pass
```

Comment:
```suggestion
    class SelfRef(BaseModel):
        x: int
        ref: Self | None

    class SubSelfRef(SelfRef):
        y: str
```

Can we make this pass? If you do `SubSelfRef(x=1, ref=SelfRef(x=2), y=3)` it should fail right?

Reply:
Ha! That does fail yes. I've modified the tests and assertions. I assume this is the correct behaviour?

Reply:
Good thought @adriangb. Yep @Youssefares, that looks like the correct behavior. Thanks for the docstrings in the tests!